tests/gradients/test_parameter_shift.py

```diff
@@ -263,6 +263,36 @@ def test_independent_parameters_analytic(self):
         assert np.allclose(j1, [exp, 0])
         assert np.allclose(j2, [0, exp])
 
+    def test_grad_recipe_parameter_dependent(self):
+        """Test that an operation with a gradient recipe that depends on
+        its instantiated parameter values works correctly within the parameter
+        shift rule"""
+
+        class RX(qml.RX):
+            @property
+            def grad_recipe(self):
+                # The gradient is given by [f(2x) - f(0)] / (2 sin(x)), by subsituting
+                # shift = x into the two term parameter-shift rule.
+                x = self.data[0]

```

Comment:
Not really relevant to this PR, but is the `data` attribute for `Operator` documented anywhere?

Reply:
Historically, `self.data` was a 'private' attribute of the operations, that stored the raw input parameter data. In initial versions of PennyLane, this raw parameter data would be `Variable` objects due to certain bookkeeping we were doing.

```pycon
>>> op.data
[Variable(0), Variable(1)]
>>> op.data[0].val
0.432
```

We also had a separate `self.parameters` property, which would inspect the `Variable` objects, and return the numerical data contained within:
```pycon
>>> op.parameters  # wrapper around op.data
[0.432, -0.654]
```

Several refactors later, the `Variable` class is long since removed, and... `self.data` and `self.parameters` are now, for all intents and purposes, identical.

Removing this redundancy is is part of the overall 'rethink the operators' discussion coming up 🙂